book/src/rfcs/3-lookup.md

```diff
@@ -0,0 +1,291 @@
+
+### Querying the table
+
+The plookup paper handles a vector of lookups $f$ which we do not have. So the first step it to create such a table from the witness columns (or registers). To do this, we define the following objects:
+
+* a **query** tells us what registers, in what order, and scaled by how much, are part of a query
+* a **query selector** tells us which rows are using the query. It is pretty much the same as a [gate selector]().
+
+Let's go over the first item in this section.
+
+For example, the following **query** tells us that we want to check if $r_0 \oplus r_2 = 2r_1$
+
+|   l   |   r   |   o   |
+| :---: | :---: | :---: |
+| 1, r0 | 1, r2 | 2, r1 |
+
+The grand product argument for the lookup consraint will look like this at this point:
+
+$$
+acc_i = acc_{i-1} \cdot \frac{\color{green}{(1+\beta)(\gamma + w_0(g^i) + j \cdot w_2(g^i) + j^2 \cdot 2 \cdot w_1(g^i))}(\gamma(1 + \beta) + t_{i-1} + \beta t_i)}{(\gamma(1+\beta) + s_{i-1} + \beta s_{i})}
+$$
+
+Not all rows need to perform queries into a lookup table. We will use a query selector in the next section to make the constraints work with this in mind.
+
+### Query selector
+
+The associated **query selector** tells us on which rows the query into the XOR lookup table occurs.
+
+|  row  | query selector |
+| :---: | :------------: |
+|   0   |       1        |
+|   1   |       0        |
+
+
+Both the (XOR) lookup table and the query are built-ins in kimchi. The query selector is derived from the circuit at setup time. Currently only the ChaCha gates make use of the lookups.
+
+The grand product argument for the lookup constraint looks like this now:
+
+$$
+acc_i = acc_{i-1} \cdot \frac{\color{green}{(1+\beta) \cdot query} \cdot (\gamma(1 + \beta) + t_{i-1} + \beta t_i)}{(\gamma(1+\beta) + s_{i-1} + \beta s_{i})}
+$$
+
+where $\color{green}{query}$ is constructed so that a dummy query ($0 \oplus 0 = 0$) is used on rows that don't have a query.
+
+$$
+\begin{align}
+query = &\ selector \cdot (\gamma + w_0(g^i) + j \cdot w_2(g^i) + j^2 \cdot 2 \cdot w_1(g^i)) + \\
+&\ (1- selector) \cdot (\gamma + 0 + j \cdot 0 + j^2 \cdot 0)
+\end{align}
+$$
+
+### Queries, not query
+
+Since we allow multiple queries per row, we define multiple **queries**, where each query is associated with a **lookup selector**. 
+
+At the moment of this writing, the `ChaCha` gates all perform $4$ queries in a row. Thus, $4$ is trivially the largest number of queries that happen in a row.
+
+**Important**: to make constraints work, this means that each row must make 4 queries. (Potentially some or all of them are dummy queries.)
+
+For example, the `ChaCha0`, `ChaCha1`, and `ChaCha2` gates will apply the following 4 XOR queries on the current and following rows:
+
+|   l   |   r   |   o    | -   |   l   |   r   |   o    | -   |   l   |   r   |   o    | -   |   l   |   r    |   o    |
```

Comment:
This part with the table with 15 columns is a bit obscure to me in the sense that it is coincidentally the same number of kimchi columns so for me it was misleading. As I commented on the hackmd:

>  My understanding was that the XORs in the quarter rounds were checked against the lookup table with all of the XOR 4-bit entries but eventually they had to be linked to some witness cells in the rows somehow, but the way does not click in my head yet.

Reply:
I answered there, but if you have an idea on how I can make this part clearer let me know :o I'll think about it